system/jlib/jfile.cpp

```diff
@@ -760,6 +760,79 @@ bool CFile::remove()
 #endif
 }
 
+static std::atomic<unsigned> renameRetries{(unsigned)-1}; // initially uninitialized, set on 1st use (in rename)
+
+void setRenameRetries(unsigned numRetries)
+{
+    renameRetries = numRetries;
+}
+
+static unsigned getRenameRetries()
+{
+    unsigned retries;
+    // NB: potentially could be >1 thread here, but that's ok.
+    try
+    {
+        retries = getGlobalConfigSP()->getPropInt("expert/@numRenameRetries");
+    }
+    catch (IException *e) // handle cases where config. not available
+    {
+        EXCLOG(e, "doRename");
+        e->Release();
+        retries = 0;
+    }
+    dbgassertex((unsigned)-1 != retries);
+    setRenameRetries(retries);
+    return retries;
+}
+
+static void doRename(const char *src, const char *dst, const char *callerName)
+{
+    unsigned retriesLeft = renameRetries;
+    unsigned retry = 0;
+    while (true)
+    {
+        if (-1 != ::rename(src, dst))
+            return; // success
+
+        if ((unsigned)-1 == retriesLeft)
+            retriesLeft = getRenameRetries();
+
+        int err = errno; // preserve to use after tracing if exhausted retry attempts
+
+#ifndef _WIN32 // NB: this is primary here to help track down the issue reported in HPCC-28454
+        // why did rename fail? - check that src and path of dst exists (and/or is accessible), and trace it's gid and uid, size and is_dir status.
+        struct stat srcInfo;
+        if (0 != stat(src, &srcInfo))
```

Comment:
If the rename() fails because the file does not exist yet (I know that's just a guess at what might be happening) then perhaps we should not throw upon stat() failure ?  Perhaps just report stat() failure ??

Reply:
on reflection I agree - the theory that this issue is being caused because the file is still just one of the theories, but it is probably the most likely. As it stands it will prove that, but yes it would be better to avoid the throw here if it doesn't and still allow it to retry. I'll change it.